commit 521205a0268bcd8e358221439498ba88f22fe879 (refs/changes/35/3835/1)
Author: Cody Peter Mello <cody.mello@joyent.com>
Date:   2018-04-17T17:23:38+00:00 (1 year, 6 months ago)
    
    TRITON-183 Server NICs in NAPI cannot have "underlay" updated to "true"

diff --git a/lib/models/nic/common.js b/lib/models/nic/common.js
index 71f4c7e..4ec7a2a 100644
--- a/lib/models/nic/common.js
+++ b/lib/models/nic/common.js
@@ -427,7 +427,7 @@ function validateUnderlayServer(opts, params, _parsedParams, callback) {
     if (params.belongs_to_type !== undefined) {
         belongs_to_type = params.belongs_to_type;
     } else {
-        belongs_to_type = opts.existingNic.belongs_to_type;
+        belongs_to_type = opts.existingNic.params.belongs_to_type;
     }
     if (params.underlay && belongs_to_type !== 'server') {
         callback(errors.invalidParam('underlay',
diff --git a/lib/models/nic/update.js b/lib/models/nic/update.js
index 0688761..a45e7eb 100644
--- a/lib/models/nic/update.js
+++ b/lib/models/nic/update.js
@@ -55,6 +55,7 @@ var UPDATE_PARAMS = [
     'primary',
     'reserved',
     'state',
+    'underlay',
     'vlan_id'
 ];
 
diff --git a/test/integration/fabrics.test.js b/test/integration/fabrics.test.js
index cc2f31f..4339e50 100644
--- a/test/integration/fabrics.test.js
+++ b/test/integration/fabrics.test.js
@@ -859,7 +859,8 @@ test('provision server nics', function (t) {
         });
     });
 
-    t.test('REAL_NETS[0]: fail to provision underlay NIC', function (t2) {
+    t.test('REAL_NETS[0]: fail to provision underlay NIC (bad belongs_to_type)',
+        function (t2) {
         mod_nic.provision(t2, {
             fillInMissing: true,
             net: REAL_NETS[0].uuid,
@@ -951,6 +952,46 @@ test('provision server nics', function (t) {
 });
 
 
+test('Convert normal server NIC into underlay NIC', function (t) {
+    var server = mod_uuid.v4();
+    var nic;
+
+    t.test('Provision server NIC', function (t2) {
+        mod_nic.provision(t2, {
+            fillInMissing: true,
+            net: REAL_NETS[0].uuid,
+            params: {
+                belongs_to_type: 'server',
+                belongs_to_uuid: server,
+                owner_uuid: ADMIN_OWNER
+            },
+            exp: mod_net.addNetParams(REAL_NETS[0], {
+                belongs_to_type: 'server',
+                belongs_to_uuid: server,
+                owner_uuid: ADMIN_OWNER
+            })
+        });
+    });
+
+    t.test('Update server NIC to underlay=true', function (t2) {
+        nic = mod_nic.lastCreated();
+        t.ok(nic, 'have last created nic');
+        t.ok(!nic.underlay, 'nic is not underlay=true');
+
+        mod_nic.update(t2, {
+            mac: nic.mac,
+            ignore: [ 'modified_timestamp' ],
+            params: {
+                underlay: true
+            },
+            exp: mod_jsprim.mergeObjects(nic, {
+                underlay: true
+            })
+        });
+    });
+});
+
+
 test('provision zone nics', function (t) {
 
     var nicTags = [
